src/settings.c

```diff
@@ -0,0 +1,273 @@
+
+G_DEFINE_TYPE_WITH_CODE (Settings, settings, XDP_TYPE_SETTINGS_SKELETON,
+                         G_IMPLEMENT_INTERFACE (XDP_TYPE_SETTINGS, settings_iface_init));
+
+typedef struct {
+  GSettingsSchema *schema;
+  GSettings *settings;
+} SettingsBundle;
+
+static SettingsBundle *
+settings_bundle_new (GSettingsSchema *schema,
+                     GSettings       *settings)
+{
+  SettingsBundle *bundle = g_new (SettingsBundle, 1);
+  bundle->schema = schema;
+  bundle->settings = settings;
+  return bundle;
+}
+
+static void
+settings_bundle_free (SettingsBundle *bundle)
+{
+  g_object_unref (bundle->schema);
+  g_object_unref (bundle->settings);
+  g_free (bundle);
+}
+
+static gboolean
+namespace_matches (const char *namespace,
+                   const char *pattern)
+{
+  size_t pattern_len;
+
+  if (pattern[0] == '\0')
+    return TRUE;
+  if (strcmp (namespace, pattern) == 0)
+    return TRUE;
+
+  pattern_len = strlen (pattern);
+  if (pattern[pattern_len - 1] == '*' && strncmp (namespace, pattern, pattern_len - 1) == 0)
+    return TRUE;
+
+  return FALSE;
+}
+
+static gboolean
+settings_handle_read_all (XdpSettings           *object,
+                          GDBusMethodInvocation *invocation,
+                          const char            *arg_namespace)
+{
+  Settings *self = (Settings *)object;
+  GVariantBuilder builder = G_VARIANT_BUILDER_INIT (G_VARIANT_TYPE ("(a{sa{sv}})"));
+  GHashTableIter iter;
+  char *key;
+  SettingsBundle *value;
+
+  g_variant_builder_open (&builder, G_VARIANT_TYPE ("a{sa{sv}}"));
+  g_hash_table_iter_init (&iter, self->settings);
+  while (g_hash_table_iter_next (&iter, (gpointer *)&key, (gpointer *)&value))
+    {
+      g_auto (GStrv) keys = g_settings_schema_list_keys (value->schema);
+      GVariantDict dict;
+      gsize i;
+
+      if (!namespace_matches (key, arg_namespace))
+        continue;
+
+      g_variant_dict_init (&dict, NULL);
+      for (i = 0; keys[i]; ++i)
+        g_variant_dict_insert_value (&dict, keys[i], g_settings_get_value (value->settings, keys[i]));
+
+      g_variant_builder_add (&builder, "{s@a{sv}}", key, g_variant_dict_end (&dict));
+    }
+  g_variant_builder_close (&builder);
+
+  g_dbus_method_invocation_return_value (invocation, g_variant_builder_end (&builder));
+  return TRUE;
+}
+
+static gboolean
+settings_handle_read (XdpSettings           *object,
+                      GDBusMethodInvocation *invocation,
+                      const char            *arg_namespace,
+                      const char            *arg_key)
+{
+  Settings *self = (Settings *)object;
+
+  g_debug ("Read %s %s", arg_namespace, arg_key);
+
+  // TODO: Handle kdeglobals via same interface
+  if (!g_hash_table_contains (self->settings, arg_namespace))
+    {
+      g_debug ("Attempted to read from unknown namespace %s", arg_namespace);
+      g_dbus_method_invocation_return_error_literal (invocation, G_IO_ERROR, G_IO_ERROR_FAILED,
```

Comment:
I think we should be using FLATPAL_PORTAL_ERROR here

Reply:
`XDG_DESKTOP_PORTAL_ERROR` you probably mean? Done.